apps/meteor/imports/message-read-receipt/server/lib/ReadReceipt.js

```diff
@@ -47,11 +47,29 @@ export const ReadReceipt = {
 			return;
 		}
 
-		this.storeReadReceipts(Messages.findUnreadMessagesByRoomAndDate(roomId, userLastSeen), roomId, userId);
+		this.storeReadReceipts(Messages.findVisibleUnreadMessagesByRoomAndDate(roomId, userLastSeen), roomId, userId);
 
 		updateMessages(room);
 	},
 
+	markThreadMessagesAsRead(message, roomId, userId) {
```

Comment:
we need to find a better way to do this.. I can look into some real data to give you more context, but I've seen threads with a few hundreds of messages, which would cause some trouble here.

there are some minor optimizations missing, like returning only `_id` from `findVisibleThreadByThreadId`, but I don't think that would be enough.

I think we come up with a good solution for this, but we'd probably need a new collection and a few days to work on it.